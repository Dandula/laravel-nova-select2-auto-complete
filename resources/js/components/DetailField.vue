<template>
    <div class="flex border-b border-40">
        <div class="w-1/4 py-4">
            <slot>
                <h4 class="font-normal text-80">
                    {{ field.name }}
                </h4>
            </slot>
        </div>
        <div class="w-3/4 py-4">
            <slot name="value">
                <div v-if="field.value">
                    <div v-for="option in field.options">
                        <template v-if="field.showAsLink">
                            <router-link
                                v-if="valueExist(field.value, option.id)"
                                :to="{
                                        name: 'detail',
                                        params: {
                                            resourceName: field.linkToResource || resourceName,
                                            resourceId: option.id
                                        }
                                    }"
                                class="no-underline dim text-primary font-bold"
                            >
                                {{ labelFor(option.text) }}
                            </router-link>
                        </template>
                        <template v-else>
                            <p v-if="valueExist(field.value, option.id)" class="text-left" v-html="labelFor(option.text)"></p>
                        </template>
                    </div>
                </div>
                <p v-else>&mdash;</p>
            </slot>
        </div>
    </div>
</template>

<script>
    export default {
        props: ['resource', 'resourceName', 'resourceId', 'field'],
        mounted() {
            this.$ga.page('/detail')
        },
        methods: {
            labelFor(text) {
                return text.replace(/(?:\r\n|\r|\n)/g, ' ')
            },
            valueExist(fieldValue, optionValue) {
                fieldValue = Array.isArray(fieldValue) ? fieldValue : [fieldValue]

                return fieldValue.indexOf(optionValue) !== -1
            }
        }
    }
</script>
