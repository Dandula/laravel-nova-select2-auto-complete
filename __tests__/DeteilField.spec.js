import {mount, shallowMount, createLocalVue} from '@vue/test-utils'
import PortalVue from 'portal-vue'
import DetailField from '../resources/js/components/DetailField'

const localVue = createLocalVue()
localVue.use(PortalVue)

// shallowMount(DetailField, {
//     stubs: ['router-link', 'router-view']
// })

describe('DetailField', () => {
    test('it renders correctly without field.value', () => {
        const wrapper = mount(DetailField, {
            localVue,
            propsData: {
                resourceName: 'category',
                field: {
                    name: 'Parent category',
                    value: null
                }
            }
        })

        expect(wrapper).toMatchSnapshot()
    })

    test('it field.value is null', () => {
        const wrapper = mount(DetailField, {
            localVue,
            propsData: {
                resourceName: 'category',
                field: {
                    name: 'Parent category',
                    value: null
                }
            }
        })

        expect(wrapper.props().field.value).toBe(null)
    })

    test('it field name is Parent category', () => {
        const wrapper = mount(DetailField, {
            localVue,
            propsData: {
                resourceName: 'category',
                field: {
                    name: 'Parent category',
                    value: null
                }
            }
        })

        expect(wrapper.text()).toContain('Parent category')
    })

    test('it renders correctly with field.value + options', () => {
        const wrapper = mount(DetailField, {
            localVue,
            propsData: {
                resourceName: 'category',
                field: {
                    name: 'Parent category',
                    value: 3,
                    options: [
                        {id: 1, text: 'One'},
                        {id: 2, text: 'Two'},
                        {id: 3, text: 'Three'}
                    ]
                }
            }
        })

        expect(wrapper).toMatchSnapshot()
    })

    test('it field.value is 3', () => {
        const wrapper = mount(DetailField, {
            localVue,
            propsData: {
                resourceName: 'category',
                field: {
                    name: 'Parent category',
                    value: 3,
                    options: [
                        {id: 1, text: 'One'},
                        {id: 2, text: 'Two'},
                        {id: 3, text: 'Three'}
                    ]
                }
            }
        })

        expect(wrapper.props().field.value).toBe(3)
    })

    test('it renders correctly with field.value + options + showAsLink', () => {
        const wrapper = mount(DetailField, {
            localVue,
            propsData: {
                resourceName: 'category',
                field: {
                    name: 'Parent category',
                    value: 3,
                    options: [
                        {id: 1, text: 'One'},
                        {id: 2, text: 'Two'},
                        {id: 3, text: 'Three'}
                    ],
                    showAsLink: true
                }
            },
            stubs: ['router-link', 'router-view']
        })

        expect(wrapper).toMatchSnapshot()
    })
})
