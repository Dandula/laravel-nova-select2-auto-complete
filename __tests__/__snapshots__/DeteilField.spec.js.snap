// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailField it renders correctly with field.value + options + showAsLink 1`] = `
<div class="flex border-b border-40">
  <div class="w-1/4 py-4">
    <h4 class="font-normal text-80">
      Parent category
    </h4>
  </div>
  <div class="w-3/4 py-4">
    <div>
      <div>
        <!---->
      </div>
      <div>
        <!---->
      </div>
      <div>
        <router-link-stub to="[object Object]" class="no-underline dim text-primary font-bold">
          Three
        </router-link-stub>
      </div>
    </div>
  </div>
</div>
`;

exports[`DetailField it renders correctly with field.value + options 1`] = `
<div class="flex border-b border-40">
  <div class="w-1/4 py-4">
    <h4 class="font-normal text-80">
      Parent category
    </h4>
  </div>
  <div class="w-3/4 py-4">
    <div>
      <div>
        <!---->
      </div>
      <div>
        <!---->
      </div>
      <div>
        <p class="text-left">Three</p>
      </div>
    </div>
  </div>
</div>
`;

exports[`DetailField it renders correctly without field.value 1`] = `
<div class="flex border-b border-40">
  <div class="w-1/4 py-4">
    <h4 class="font-normal text-80">
      Parent category
    </h4>
  </div>
  <div class="w-3/4 py-4">
    <p>â€”</p>
  </div>
</div>
`;
